FinniversKit/Sources/Components/TextField/SwiftUI/FinnTextField.swift

```diff
@@ -0,0 +1,266 @@
+                onEditingChanged: { editing in updateStyle(editing ? .focused : .default) },
+                onCommit: { evaluateTextState() }
+            )
+            .fixedSize(horizontal: false, vertical: true)
+            .padding(.vertical, .spacingS + .spacingXS)
+            .padding(.horizontal, .spacingS)
+            .background(Color.bgSecondary)
+            .overlay(underline, alignment: .bottom)
+            .overlay(textfieldTrailingButton.padding(.trailing, .spacingXS), alignment: .trailing)
+
+            makeHelpTextLabel()
+                .transition(.asymmetricSlide)
+                .padding(.top, 0)
+                .padding(.bottom, 1)
+        }
+    }
+
+    private var underline: some View {
+        Rectangle()
+            .frame(height: style.underlineHeight)
+            .foregroundColor(style.underlineColor)
+    }
+
+    @ViewBuilder
+    private func makeHelpTextLabel() -> some View {
+        let label = Text(helpText ?? "")
+            .finnFont(.detail)
+            .foregroundColor(style.helpTextColor)
+
+        if shouldShowHelpText {
+            label
+        } else {
+            label.hidden()
+        }
+    }
+
+    @ViewBuilder
+    var textfieldTrailingButton: some View {
+        switch input {
+        case .default:
+            if style == .focused {
+                clearButton
+            }
+        case .secure:
+            disclosePasswordButton
+        default:
+            SwiftUI.EmptyView()
+        }
+    }
+
+    private var clearButton: some View {
+        SwiftUI.Button(action: {
+            text = ""
+        }, label: {
+            Image(.remove)
+                .renderingMode(.template)
+                .foregroundColor(Color.iconPrimary)
+        })
+    }
+
+    private var disclosePasswordButton: some View {
+        SwiftUI.Button(action: {
+            disclosePassword.toggle()
+        }, label: {
+            Image(.view)
+                .renderingMode(.template)
+                .foregroundColor(disclosePassword ? Color.accentSecondaryBlue : Color.iconPrimary)
+        })
+    }
+}
+
+@available(iOS 13.0, *)
+extension FinnTextField {
+
+    private static let emailRegex: String = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
+    private static let phoneRegex: String = "^(?:\\s*\\d){8,11}$"
+
+    private var isValidInput: Bool {
+        let isValid: Bool
+
+        switch input {
+        case .secure:
+            isValid = isValidPassword(text)
+        case .email:
+            isValid = isValidEmail(text)
+        case .phone:
+            isValid = isValidPhoneNumber(text)
+        default:
+            isValid = true
+        }
+
+        if isValid, let validator = validator {
+            return validator(text)
+        }
+
+        return isValid
+    }
+
+    private func evaluateTextState() {
+        if !isValidInput {
+            updateStyle(.error)
+        } else {
+            updateStyle(.default)
+        }
```

Comment:
I won't comment on the fact that you're checking if-not-else. 😜

Have you considered: `updateStyle(isValidInput ? .default : .error)`

Reply:
Will do!